include/bluetooth/hci.h

```diff
@@ -1914,6 +1915,28 @@ int bt_hci_cmd_send(u16_t opcode, struct net_buf *buf);
 int bt_hci_cmd_send_sync(u16_t opcode, struct net_buf *buf,
 			 struct net_buf **rsp);
 
+/** @typedef bt_hci_vnd_evt_cb_t
+  * @brief Callback type for vendor handling of HCI Vendor-Specific Events.
+  *
+  * A function of this type is registered with bt_hci_register_vnd_evt_cb()
+  * and will be called for any HCI Vendor-Specific Event. If the function
+  * can decode and handle the event it must return 0, otherwise -EINVAL.
+  *
+  * @param buf Buffer containing event parameters.
+  *
+  * @return 0 on success or -EINVAL value on failure to handle.
+  */
+typedef int bt_hci_vnd_evt_cb_t(struct net_buf_simple *buf);
+
+/** Register user callback for HCI Vendor-Specific Events
+  *
+  * @param cb Callback to be called when the stack receives a
+  *           HCI Vendor-Specific Event.
+  *
+  * @return 0 on success or negative error value on failure.
+  */
+int bt_hci_register_vnd_evt_cb(bt_hci_vnd_evt_cb_t cb);
```

Comment:
This is a compromise between cleanliness and optimization, but another option would be to use a weak function here. We did this here: 
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/bluetooth/hci/h4.c#L428

This is perhaps less clean from an architectural point of view, but saves RAM and ROM.

Thoughts?

Reply:
I don't like weak functions simply because of compiler compatibility, so I would prefer a callback. If we need to save RAM and ROM we should probably do that with ifdefs instead.
Also we usually use callbacks, so I'd prefer to be consistent.